services/device-registry-file/src/main/java/org/eclipse/hono/deviceregistry/file/FileBasedRegistrationService.java

```diff
@@ -520,6 +527,120 @@ private JsonObject convertDevice(final String deviceId, final Device payload) {
         return identities.computeIfAbsent(tenantId, id -> new ConcurrentHashMap<>());
     }
 
+    /**
+     * Return all devices for a tenant.
+     *
+     * The order of devices cannot be guaranteed by the underlying in-memory storage
+     * so the paging may be inconsistent.
+     */
+    @Override
+    public Future<OperationResult<SearchDevicesResult>> searchDevices(final String tenantId, final int pageSize, final int pageOffset,
+                                                                       final List<Filter> filters, final List<Sort> sortOptions, final Span span) {
+
+        final Set allDevicesIds = getDevicesForTenant(tenantId).keySet();
+        List<DeviceWithId> devicesList = new ArrayList<>();
+
+        for (int i = 0; i < allDevicesIds.size(); i++) {
+            final String id = (String) allDevicesIds.toArray()[i];
+            final Device device = getDevicesForTenant(tenantId).get(id).getValue();
+            final DeviceWithId withId = DeviceWithId.from(id, device);
+
+            devicesList.add(withId);
+        }
+
+        final List<Predicate<DeviceWithId>> predicates = new ArrayList<>();
+        for (Filter f : filters) {
+            predicates.add(buildJsonBasedPredicate(f));
+        }
+
+        //Regroup predicates in a single one with AND.
+        final Predicate<DeviceWithId> predicate = predicates.stream()
+                .reduce(x -> true, Predicate::and);
+
+        // 1. apply the filter options to the list
+        devicesList = devicesList
+                .stream()
+                .filter(predicate)
+                .collect(Collectors.toList());
+
+        // 2. sort the selected elements
+        for (Sort s : sortOptions) {
+            devicesList = sortDeviceList(devicesList, s);
+        }
+
+        // 3. limit the results according to the page settings.
+        final List<DeviceWithId> returnDevicesList = new ArrayList<>();
+
+        for (int i = pageOffset; i < devicesList.size() && returnDevicesList.size() < pageSize; i++) {
+            returnDevicesList.add(devicesList.get(i));
+        }
+
+        return Future.succeededFuture(
```

Comment:
As per the spec, if none of the devices are matched, then 404 should be returned.